wagtailinventory/views.py

```diff
@@ -1,35 +1,65 @@
-from django.core.paginator import Paginator
-from django.http import HttpResponseBadRequest
-from django.shortcuts import render
-from django.views.generic import View
+from django import forms
 
+from wagtail.admin.filters import WagtailFilterSet
+from wagtail.admin.views.reports import PageReportView
 from wagtail.models import Page
 
-from wagtailinventory.forms import PageBlockQueryFormSet
+import django_filters
 
+from wagtailinventory.checks import dal_select2_check_all
+from wagtailinventory.models import PageBlock
 
-class SearchView(View):
-    template_name = "wagtailinventory/search.html"
 
-    def get(self, request):
-        if len(request.GET) > 1:
-            formset = PageBlockQueryFormSet(request.GET)
-            if not formset.is_valid():
-                return HttpResponseBadRequest("invalid query")
+# If django-autocomplete-light is present, we'll do two things:
+#
+# 1. Set the widget used for the block include/exclude filters to DAL's
+#    multi-select;
+# 2. Create a list view populated with available blocks to autocomplete.
+#
+# If it's not present, we'll fall back on Django's built-in SelectMultiple
+# widget.
+if not dal_select2_check_all():  # pragma: no cover
+    autocomplete = None
+    block_widget = forms.SelectMultiple()
+else:
+    from dal import autocomplete
 
-            queryset = formset.get_query()
-        else:
-            formset = PageBlockQueryFormSet()
-            queryset = Page.objects.all()
+    block_widget = autocomplete.Select2Multiple(url="block-autocomplete")
 
-        queryset = queryset.order_by("title")
+    class BlockAutocompleteView(autocomplete.Select2ListView):
+        def get_list(self):
+            return (
+                PageBlock.objects.distinct()
+                .order_by("block")
+                .values_list("block", flat=True)
+            )
 
-        paginator = Paginator(queryset, per_page=20)
-        pages = paginator.get_page(request.GET.get("p"))
 
-        for page in pages:
-            page.can_choose = True
+class BlockInventoryFilterSet(WagtailFilterSet):
+    include_page_blocks = django_filters.AllValuesMultipleFilter(
+        field_name="page_blocks__block",
+        label="Include Blocks",
+        distinct=True,
+        widget=block_widget,
+    )
+    exclude_page_blocks = django_filters.AllValuesMultipleFilter(
```

Comment:
Using the fallback block widget, the include/exclude selects include a "None" option which results in an error when selected:

<img width="353" alt="image" src="https://user-images.githubusercontent.com/654645/223153270-60e4c6a8-9415-4800-b350-5e9a05cfd453.png">

There aren't any PageBlocks with a null `block` value.

As a user, I would expect that choosing "None" for "include blocks" would return pages without any blocks defined; similarly, choosing "None" for "exclude blocks" would return only pages with blocks defined.

Reply:
I'm looking into how to make this better, but this seems to come from [django-filters' `null` values for `ChoiceFilter`s](https://django-filter.readthedocs.io/en/stable/ref/filters.html#modelchoicefilter-and-modelmultiplechoicefilter-arguments). Right off I'm not sure I see a way to change this behavior without additional complexity.

Reply:
So, it looks like this is an issue where `ChoiceFilter` offers a null value but `AllValuesMultipleFilter` can't validate the null value is valid and errors before filtering. This is a bug in django-filters.

Rather than propose to fix this upstream (and wait for Wagtail to bring it in), or to fix it in wagtail-inventory (which adds a lot of complexity), I’m dropping no-DAL as an option.

From [fb15615](https://github.com/cfpb/wagtail-inventory/pull/68/commits/fb156159645accf72313448e519a96f4172fb872), django-autocomplete-light is required. This results in less overall code, and less complexity.